# 09. Advanced NumPy Operations
## 📚 Learning Objectives

By completing this notebook, you will:
- Master advanced NumPy operations (broadcasting, vectorization)
- Understand matrix operations and linear algebra with NumPy
- Apply advanced indexing and slicing techniques
- Optimize numerical computations using NumPy

## 🔗 Where this fits

**Builds on:** Course 03 (AIAT 113) — Unit 1, lessons 01-02 (vectors, matrices and matrix operations) — the same algebra, expressed as broadcasting and vectorised NumPy.

**Used later in:** Course 05 — Unit 5, lessons 06-07, where vectorisation is the difference between a workable and an unworkable dataset size.

---


## 🎯 The case: a search engine that is one matrix multiply, repeated

In 1998 Larry Page and Sergey Brin published *"The Anatomy of a Large-Scale Hypertextual
Web Search Engine"*. The ranking idea at its centre, **PageRank**, is not a program full
of special cases. It is a single line of linear algebra applied over and over:
`x ← M̂ x` — multiply the link matrix by the current score vector, repeat until it stops
changing. In the companion technical report (Page, Brin, Motwani & Winograd, *The PageRank
Citation Ranking*, Stanford InfoLab, 1998) they report that "PageRank on a large 322
million link database converges to a reasonable tolerance in roughly **52 iterations**". The operation you write below as `A @ B` is that operation.

This is what "advanced NumPy" buys you: an algorithm whose *description* and whose *code*
are the same length. Broadcasting, matrix multiplication and boolean indexing let you write
mathematics instead of loops — and the mathematics is what someone else can check.

**What goes wrong without it.** You already have the number: further down this notebook,
converting 1,000,000 real CIC-IDS2017 flow durations from microseconds to seconds takes
**0.00073 s** vectorized and **0.06663 s** in a Python loop — **91.7×**, on identical
input, for identical output. Now imagine that loop inside PageRank's 52 iterations over 322
million links. Vectorization is not a style preference at that scale; it is the difference
between a search engine and a research proposal.


## The Story

**BEFORE**: You know basic NumPy but don't know advanced operations for efficient data processing.

**AFTER**: You'll master advanced NumPy: broadcasting, vectorization, array manipulation, and performance optimization!

**Why this matters**: Advanced NumPy Operations is essential for building complete, professional data science solutions!

---

**All concepts are explained in the code comments below - you can learn everything from this notebook alone!**

---

## 🔗 Building on Example 2

**Remember from Example 2?**
- We learned basic NumPy arrays and operations
- We saw how NumPy is faster than Python lists
- But we only scratched the surface!

**This notebook goes deeper!**
- We'll learn **broadcasting** - powerful array operations
- We'll learn **vectorization** - fast operations on entire arrays
- We'll learn **matrix operations** - linear algebra with NumPy
- We'll learn **advanced indexing** - powerful data selection

**This unlocks the full power of NumPy!**

---

## Learning Objectives
1. Understand broadcasting and vectorization
2. Perform matrix operations (multiplication, transpose, inverse)
3. Use advanced indexing (boolean, fancy indexing)
4. Optimize computations using NumPy best practices

## 📥 Inputs & 📤 Outputs

**Inputs:** What we use in this notebook

- Advanced array operations, building on Example 2
- NumPy, matplotlib

**Outputs:** What you'll see when you run the cells

- Numerical results
- Visualizations

---

In [1]:
# WHAT: Import NumPy and preview the four topics of this notebook.
# WHY: Broadcasting, vectorization, linear algebra, and advanced indexing are the tools that replace slow Python loops.

# Step 1: Import NumPy
import numpy as np
import time

print("✅ NumPy imported successfully!")
print("\n📚 This notebook covers:")
print("   - Broadcasting (automatic dimension expansion)")
print("   - Vectorization (fast array operations)")
print("   - Matrix operations (linear algebra)")
print("   - Advanced indexing (powerful selection)")

✅ NumPy imported successfully!

📚 This notebook covers:
   - Broadcasting (automatic dimension expansion)
   - Vectorization (fast array operations)
   - Matrix operations (linear algebra)
   - Advanced indexing (powerful selection)


In [2]:
# Step 2: Broadcasting - Automatic Dimension Expansion
print("=" * 70)
print("=" * 70)

# Example: Add scalar to array (broadcasting)
arr = np.array([1, 2, 3, 4, 5])
result = arr + 10  # Broadcasting: 10 is added to each element
print(f"Array: {arr}")
print(f"Array + 10: {result}")
print("\n💡 Broadcasting automatically expands dimensions!")

# Example: Add 1D array to 2D array
arr_2d = np.array([[1, 2, 3], [4, 5, 6]])
arr_1d = np.array([10, 20, 30])
result_2d = arr_2d + arr_1d  # Broadcasting: adds to each row
print(f"\n2D Array:\n{arr_2d}")
print(f"1D Array: {arr_1d}")
print(f"Result (broadcasted):\n{result_2d}")

Array: [1 2 3 4 5]
Array + 10: [11 12 13 14 15]

💡 Broadcasting automatically expands dimensions!

2D Array:
[[1 2 3]
 [4 5 6]]
1D Array: [10 20 30]
Result (broadcasted):
[[11 22 33]
 [14 25 36]]


In [3]:
# Step 3: Matrix Operations
print("\n" + "=" * 70)
print("=" * 70)

# Create matrices
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])

print(f"Matrix A:\n{A}")
print(f"\nMatrix B:\n{B}")

# Matrix multiplication
C = np.dot(A, B)  # or A @ B
print(f"\nA @ B (matrix multiplication):\n{C}")

# Transpose
A_T = A.T
print(f"\nA transpose:\n{A_T}")

# Determinant
det_A = np.linalg.det(A)
print(f"\nDeterminant of A: {det_A:.2f}")


Matrix A:
[[1 2]
 [3 4]]

Matrix B:
[[5 6]
 [7 8]]

A @ B (matrix multiplication):
[[19 22]
 [43 50]]

A transpose:
[[1 3]
 [2 4]]

Determinant of A: -2.00


In [4]:
# Step 4: Advanced Indexing
print("\n" + "=" * 70)
print("=" * 70)

# Boolean indexing
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
mask = arr > 5  # Boolean mask
filtered = arr[mask]  # Select elements where mask is True
print(f"Array: {arr}")
print(f"Mask (arr > 5): {mask}")
print(f"Filtered (arr[mask]): {filtered}")

# Fancy indexing
indices = [0, 2, 4, 6]
selected = arr[indices]
print(f"\nFancy indexing (indices {indices}): {selected}")

# The arrays above were typed by hand so the mechanics are visible. The same two
# techniques on REAL data are how you actually select rows in daily work.
import pandas as pd
DATA_DIR = '../../../Course 04/datasets/raw/'
titanic = pd.read_csv(DATA_DIR + 'titanic.csv')
fares = titanic['Fare'].values          # a real NumPy array of 891 ticket prices

# Boolean indexing: keep only the fares above 100 pounds
expensive = fares[fares > 100]
print(f"\nReal data: {len(fares)} Titanic ticket fares")
print(f"Boolean mask (fares > 100) selects {expensive.size} of them")
print(f"   Those fares range from {expensive.min():.2f} to {expensive.max():.2f} pounds")
print(f"   The median fare overall was only {np.median(fares):.2f} - a very skewed column")

# Fancy indexing: pull out specific passengers by position
first_class_positions = np.where(titanic['Pclass'].values == 1)[0][:4]
print(f"\nFancy indexing positions {first_class_positions.tolist()} (first four 1st-class passengers):")
print(f"   Their fares: {fares[first_class_positions]}")



Array: [ 1  2  3  4  5  6  7  8  9 10]
Mask (arr > 5): [False False False False False  True  True  True  True  True]
Filtered (arr[mask]): [ 6  7  8  9 10]

Fancy indexing (indices [0, 2, 4, 6]): [1 3 5 7]

Real data: 891 Titanic ticket fares
Boolean mask (fares > 100) selects 53 of them
   Those fares range from 106.42 to 512.33 pounds
   The median fare overall was only 14.45 - a very skewed column

Fancy indexing positions [1, 3, 6, 11] (first four 1st-class passengers):
   Their fares: [71.2833 53.1    51.8625 26.55  ]


In [5]:
# WHAT: Time a vectorized conversion against the identical Python loop, on 1 million real measurements.
# WHY: The gap is the whole argument for NumPy - and it should be measured on real data, not on random noise.

print("\n" + "=" * 70)
print("4. Performance: Vectorization vs Loops")
print("=" * 70)

# 1 million real flow durations (recorded in microseconds) from CIC-IDS2017.
print("📊 Loading 1,000,000 real flow durations...")
durations = pd.read_csv(DATA_DIR + 'cicids2017.csv',
                        nrows=1_000_000, usecols=[' Flow Duration'])
large_arr = durations[' Flow Duration'].values.astype('float64')
print(f"   Loaded {large_arr.size:,} values")

# The task: convert microseconds to seconds. Same arithmetic both ways.

# Vectorized operation (fast) - the division happens in compiled C code
start = time.time()
result_vectorized = large_arr / 1e6
time_vectorized = time.time() - start

# Loop operation (slow) - one Python-level division per element
start = time.time()
result_loop = np.array([x / 1e6 for x in large_arr])
time_loop = time.time() - start

print(f"\nVectorized: {time_vectorized:.5f} seconds")
print(f"Loop: {time_loop:.5f} seconds")
print(f"\nSpeedup: {time_loop / time_vectorized:.1f}x faster with vectorization!")
print(f"Identical results: {np.allclose(result_vectorized, result_loop)}")
print(f"\nLongest flow in this sample: {result_vectorized.max():.1f} seconds")
print("\n💡 Always use vectorized operations when possible!")



4. Performance: Vectorization vs Loops
📊 Loading 1,000,000 real flow durations...


   Loaded 1,000,000 values

Vectorized: 0.00073 seconds
Loop: 0.06663 seconds

Speedup: 91.7x faster with vectorization!
Identical results: True

Longest flow in this sample: 120.0 seconds

💡 Always use vectorized operations when possible!


## ⚠️ Where this breaks

Vectorization and broadcasting are the right default, and they fail in specific,
recognisable ways:

- **Broadcasting allocates memory you did not ask for.** Two arrays of 1,000,000 elements
  shaped `(1_000_000,)` and `(1_000_000, 1)` do not combine element-wise — they broadcast
  to a 10¹² element result, roughly **8 TB** of float64. NumPy will happily try. The most
  common broadcasting bug is not a crash; it is a `MemoryError` twenty minutes into a job,
  or a silently transposed result.
- **The intermediate arrays are real.** `(a + b) * c - d` on a 1 GB array allocates
  several gigabytes of temporaries. Vectorized code trades memory for speed; when the
  memory is not there, chunking or a Numba loop (lesson 08) wins.
- **Boolean masks copy.** `arr[mask]` returns a new array, not a view. Filtering a large
  frame repeatedly in a loop quietly doubles memory each time.
- **The assumption that must hold: the operation is element-wise and order-independent.**
  Running totals, early exits, "stop when converged", and anything where element *i*
  depends on element *i−1* cannot be vectorized directly. That is exactly the territory
  lesson 08's JIT compilation was built for.
- **`np.linalg.inv` is almost never what you want.** The determinant of `A` printed above
  is −2.00, so `A` inverts fine — but for solving `Ax = b`, use `np.linalg.solve`, which is
  faster and numerically better behaved. Explicitly inverting a matrix is a classic
  textbook habit that numerical analysts spend careers warning against.
- **Cheaper alternative:** when the array does not fit, do not vectorize harder — read
  less (`usecols`, `dtype='float32'`), process in chunks, or move to Dask (Unit 5,
  lesson 02). Halving the dtype halves the memory for free, and float32 is enough
  precision for most measured data.


## 💬 Discuss

Two measured facts from the cells above: boolean indexing selected **53 of 891** Titanic
fares above £100, whose median overall is only **£14.45**; and vectorized conversion beat
the equivalent loop by **91.7×** on 1,000,000 real flow durations.

1. `fares[fares > 100]` reads like plain English and runs at C speed — but it builds a
   full boolean array of 891 values first. At what data size does that intermediate array
   start to matter, and what would you write instead? (You have a 707 MB file in this
   course to think with.)
2. Broadcasting is described here as a convenience. Give an example where broadcasting
   silently produces a *wrong shape* rather than an error — and say which is worse for a
   student: code that crashes or code that returns a plausible number of the wrong size.
3. The median fare is £14.45 and the mean is £32.20 (lesson 06 computed it). The 53
   passengers above £100 are 6% of the manifest. If you had to report *one* number about
   what a Titanic ticket cost, which do you choose, and who is harmed by the other choice?


## 📚 References

1. van der Walt, S., Colbert, S. C., & Varoquaux, G. (2011). *The NumPy Array: A Structure for Efficient Numerical Computation*. Computing in Science & Engineering, 13(2), 22-30. <https://arxiv.org/abs/1102.1523>
2. Harris, C. R., Millman, K. J., van der Walt, S. J., et al. (2020). *Array Programming with NumPy*. Nature, 585, 357-362. <https://arxiv.org/abs/2006.10256>